pinot-broker/src/main/java/org/apache/pinot/broker/api/resources/PinotClientRequest.java

```diff
@@ -141,6 +152,50 @@ public void processSqlQueryPost(String query, @Suspended AsyncResponse asyncResp
     }
   }
 
+  @DELETE
+  @Path("query/{queryId}")
+  @Produces(MediaType.APPLICATION_JSON)
+  @ApiOperation(value = "Cancel a query as identified by the queryId", notes = "No effect if no query exists for the "
+      + "given queryId on the requested broker. Query may continue to run for a short while after calling cancel as "
+      + "it's done in a non-blocking manner. The cancel method can be called multiple times.")
+  @ApiResponses(value = {
+      @ApiResponse(code = 200, message = "Success"), @ApiResponse(code = 500, message = "Internal server error"),
+      @ApiResponse(code = 404, message = "Query not found on the requested broker")
+  })
+  public String cancelQuery(
+      @ApiParam(value = "QueryId as in the format of <brokerId>_<requestId>", required = true) @PathParam("queryId")
+          String queryId,
+      @ApiParam(value = "Timeout for servers to respond the cancel request") @QueryParam("timeoutMs")
+      @DefaultValue("3000") int timeoutMs,
+      @ApiParam(value = "Return server responses for troubleshooting") @QueryParam("detailed") @DefaultValue("false")
+          boolean detailed) {
+    Map<String, Integer> serverResponses = null;
+    if (detailed) {
+      serverResponses = new HashMap<>();
+    }
+    if (!_requestHandler.cancelQuery(queryId, timeoutMs, _executor, _httpConnMgr, serverResponses)) {
+      throw new WebApplicationException(
+          Response.status(Response.Status.NOT_FOUND).entity("Query: " + queryId + " not found on the broker").build());
+    }
+    String resp = "Cancelled query: " + queryId;
+    if (detailed) {
+      resp += " with responses from servers: " + serverResponses;
+    }
+    return resp;
+  }
+
+  @GET
+  @Path("queries")
+  @Produces(MediaType.APPLICATION_JSON)
+  @ApiOperation(value = "Get queryIds of the running queries submitted via this broker", notes = "QueryIds are in the"
+      + " format of <brokerId>_<requestId>")
```

Comment:
Since the queryId is formed as `<brokerId>_<requestId>`, whichever broker receives this request, it will always return its own brokerId as the prefix. Does that mean user needs to know about the list of broker ids beforehand? If so, what's the purpose of returning the broker id in the response? Wouldn't it better to do it in controller endpoint?

Reply:
This is a good point. Broker doesn’t need to prepend its broker id to form the query id to track queries submitted to it. I carried the server side way of forming unique query id here.